### PR TITLE
[move-vm] Remove hashmap

### PR DESCRIPTION
**Pull request by runtian-zhou**
### Description

Use back btreemap to replace hashmap due to potential non-determinism concern. Also use bit vector to replace the inefficient Vec<bool>

### Test Plan
We should run a replay on this.

### CODE REVIEW

**Comment by vgao1996** [comment]
I'm a bit confused by the commit history -- it seems like you are changing a Vec<bool> into bit vec and then changing it back?

**Comment by vgao1996** [comment]
NVM I confused myself by getting the order wrong. However it seems like you are still including the bit set commits. Is that intended?

File: third_party/move/move-vm/types/src/loaded_data/runtime_types.rs [hunk 1]
```diff
@@ -194,6 +190,43 @@ pub enum Type {
     U256,
 }
 
+// Cache for the ability of struct. They will be ignored when comparing equality or Ord as they are just used for caching purpose.
+#[derive(Derivative)]
+#[derivative(Debug, Clone, Eq, Hash, PartialEq, Ord, PartialOrd)]
```
Comment (wrwg) [inline third_party/move/move-vm/types/src/loaded_data/runtime_types.rs:195]: Seems you needed this for the ignore semantics (shouldn't it have been there already for hashmap?). Not a fan of doing something like this (i.e. cheating on equality :-) ), I think the abilities belong into a different map, but this is for another time.

**Review by wrwg — APPROVED** [review]
Was the HashMap actually iterated? In any case, good to switch to ordered map.

**Review by vgao1996 — APPROVED** [review]

**Comment by runtian-zhou** [comment]
> Was the HashMap actually iterated? In any case, good to switch to ordered map.

It was iterated. The reaonsing why the current code is fine is a bit more nuanced here: we put the iteration result back to a sorted list so the happy path should be fine. The unhappy path is also not possible because there were only simple serialziation errors there. But it would be great to avoid this footgun in general.

**Pull request merged**